---
output: html_document
---

This vignette looks at the results of deconvolving HiChIP of H3K27Ac in
GM12878 and K562, starting with the results for between-replicate and
across-sample concordance, followed by differential loop and gene expression
analysis before and after deconvolution.


```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(data.table)
library(ggthemes)
library(rtracklayer)
library(RColorBrewer)
MIN_PET=3
```

```{r}
myfiles <- list.files(pattern = 'h3k27ac_results_gamma0.5_chr.*rds')
loops <- NULL
before <- NULL
after <- NULL
diff_results <- NULL
for (f in myfiles) {
  d <- readRDS(f)
  chrom <- unlist(strsplit(unlist(strsplit(f,'_'))[4],'\\.'))[1]
  if (is.null(loops)) {
    loops <- d@loops
  } else {
    loops <- rbind(loops, d@loops)
  }
  if (is.null(before)) {
    before <- d@results$before
  } else {
    before <- rbind(before, d@results$before)
  }
  if (is.null(after)) {
    after <- d@results$after
  } else {
    after <- rbind(after, d@results$after)
  }
  if (is.null(diff_results)) {
    diff_results <- d@results$df %>% mutate(chrom=chrom)
  } else {
    diff_results <- rbind(diff_results, d@results$df %>% mutate(chrom=chrom))
  }
}
```


# Between replicate and across sample concordance


```{r}
diff_results %>%
  pivot_wider(names_from = label, values_from=diff) %>%
  filter(chrom=='chr20') %>%
  ggplot(aes(x = log2(before+1), y = log2(after+1))) +
  geom_point(alpha=0.25, color='grey') +
  geom_abline(slope=1,intercept=0,lty='dashed',color='red') +
  geom_smooth(se=F, color='blue', size=0.5) +
  facet_grid(m1~m2) +
  theme_bw() +
  xlab('log2(Before+1)') +
  ylab('log2(After+1)')
ggsave('h3k27ac_scatter_pairs.pdf', height=5, width=5)
```


```{r}
diff_results %>% 
  filter((m1=='gm12878_1'&m2=='gm12878_2')|(m1=='k562_1'&m2=='k562_2')|(m1=='gm12878_1'&m2=='k562_1')|(m1=='gm12878_2'&m2=='k562_2')) %>%
  mutate(m1=toupper(m1), m2 = toupper(m2)) %>%
  pivot_wider(names_from=label, values_from=diff) %>%
  ggplot(aes(x = log2(before+1), y= log2(after+1))) +
  geom_point(color='grey',alpha=0.1) +
  geom_abline(slope=1,intercept=0, lty='dashed', color='red') +
  facet_wrap(m1~m2) +
  theme_classic() +
  xlab('log2(Before+1)') +
  ylab('log2(After+1)')
ggsave('h3k27ac_before_after_scatter.pdf', height=5, width=5)
```


# Compare gene expression of loops


Use the ENCODE RNA-seq data available on Kraken

```{r}
metadata <- fread('/data/public_data/encode_RNASeq_human/new_data/meta/human_ENCODE_RNASEQ_metadata.tsv')
metadata <- metadata %>%
  filter(grepl('GM12878', `Biosample term name`) | grepl('K562', `Biosample term name`)) %>%
  filter(Assay == 'RNA-seq')
gmat <- readMM('/data/public_data/encode_RNASeq_human/new_data/quants/paired_ended/GRCh38_tximport_gene_level_sparse_matrix.mtx')
sampleNames <- fread('/data/public_data/encode_RNASeq_human/new_data/quants/paired_ended/sampleNames.tsv',header=F)
geneNames <- fread('/data/public_data/encode_RNASeq_human/new_data/quants/paired_ended/gene_ids.tsv', header=F)
gm12878_ids <- metadata %>% filter(grepl('GM12878', `Biosample term name`)) %>% mutate(id = paste(`Experiment accession`, `File accession`, `Paired with`, sep='_')) %>% pull(id)
gm12878_idx <- which(sampleNames$V1 %in% gm12878_ids)
k562_ids <- metadata %>% filter(grepl('K562', `Biosample term name`)) %>% mutate(id = paste(`Experiment accession`, `File accession`, `Paired with`, sep='_')) %>% pull(id)
k562_idx <- which(sampleNames$V1 %in% k562_ids)
gmat <- as(gmat, 'dgCMatrix')
rownames(gmat) <- geneNames$V1
colnames(gmat) <- sampleNames$V1
```


## Annotate TSS-overlapping loops


```{r}
# get gene TSS
gencode <- import('/rafalab/lzou/resources/gencode.v42lift37.annotation.gtf.gz')
gencode <- gencode[gencode$type=='gene',]
gene_df <- data.frame(gene_name = gencode$gene_name, gene_id = gencode$gene_id, strand = strand(gencode),
                      chrom = seqnames(gencode), start = start(gencode), end = end(gencode))
gene_df <- gene_df %>%
  mutate(TSS_start = ifelse(strand == '+', start, end),
         TSS_end = ifelse(strand=='+', start+1, end-1))

gene_df$gene_id <- gsub('_.*','',gene_df$gene_id)
# overlap loop start/ends to get ones overlapping TSS
# get all the gene TSS that touch that loop
# then get the average gene expression for gm12878 and k562 for those genes

ovl <- findOverlaps(loops %>% dplyr::select(chr1, s1, e1) %>% dplyr::rename(seqnames = chr1, start = s1, end = e1) %>% makeGRangesFromDataFrame(),
             gene_df %>% dplyr::select(chrom, TSS_start, TSS_end) %>% dplyr::rename(seqnames = chrom, start = TSS_start, end = TSS_end) %>% makeGRangesFromDataFrame()) %>%
  data.frame() 
ovl$gene <- gene_df$gene_id[ovl$subjectHits]
ovl$avg_gm12878 <- NA
ovl$avg_gm12878[ovl$gene %in% geneNames$V1] <- rowMeans(gmat[ovl$gene[ovl$gene %in% geneNames$V1],gm12878_idx])
ovl$avg_k562 <- NA
ovl$avg_k562[ovl$gene %in% geneNames$V1] <- rowMeans(gmat[ovl$gene[ovl$gene %in% geneNames$V1],k562_idx])
gex <- ovl %>%
  group_by(queryHits) %>%
  summarise(gm12878 = mean(avg_gm12878, na.rm=T), k562 = mean(avg_k562, na.rm=T)) %>%
  filter(!is.na(gm12878), !is.na(k562))

loops$gm12878_gex_1 <- NA
loops$gm12878_gex_1[gex$queryHits] <- gex$gm12878
loops$k562_gex_1 <- NA
loops$k562_gex_1[gex$queryHits] <- gex$k562


ovl <- findOverlaps(loops %>% dplyr::select(chr2, s2, e2) %>% dplyr::rename(seqnames = chr2, start = s2, end = e2) %>% makeGRangesFromDataFrame(),
             gene_df %>% dplyr::select(chrom, TSS_start, TSS_end) %>% dplyr::rename(seqnames = chrom, start = TSS_start, end = TSS_end) %>% makeGRangesFromDataFrame()) %>%
  data.frame() 
ovl$gene <- gene_df$gene_id[ovl$subjectHits]
ovl$avg_gm12878 <- NA
ovl$avg_gm12878[ovl$gene %in% geneNames$V1] <- rowMeans(gmat[ovl$gene[ovl$gene %in% geneNames$V1],gm12878_idx])
ovl$avg_k562 <- NA
ovl$avg_k562[ovl$gene %in% geneNames$V1] <- rowMeans(gmat[ovl$gene[ovl$gene %in% geneNames$V1],k562_idx])
gex <- ovl %>%
  group_by(queryHits) %>%
  summarise(gm12878 = mean(avg_gm12878, na.rm=T), k562 = mean(avg_k562, na.rm=T)) %>%
  filter(!is.na(gm12878), !is.na(k562))

loops$gm12878_gex_2 <- NA
loops$gm12878_gex_2[gex$queryHits] <- gex$gm12878
loops$k562_gex_2 <- NA
loops$k562_gex_2[gex$queryHits] <- gex$k562
```


## First pass: plot average difference in loop vs. average difference in expression 

(for loops overlapping genes)


```{r}
loops$gm12878_before <- rowMeans(before[, 1:2])
loops$k562_before <- rowMeans(before[, 3:5])
loops$gm12878_after <- rowMeans(after[, 1:2])
loops$k562_after <- rowMeans(after[, 3:5])
loops$diff_before <- loops$gm12878_before - loops$k562_before
loops$diff_after <- loops$gm12878_after - loops$k562_after

loops$id <- 1:nrow(loops)
```

```{r}
loops_gex <- loops %>%
  filter(!is.na(gm12878_gex_1) | !is.na(gm12878_gex_2))
# consolidate gene expression on both sides by taking the max
loops_gex <- loops_gex %>%
  mutate(gm12878_gex = case_when(
    is.na(gm12878_gex_1) ~ gm12878_gex_2,
    is.na(gm12878_gex_2) ~ gm12878_gex_1,
    gm12878_gex_1 > gm12878_gex_2 ~ gm12878_gex_1,
    gm12878_gex_2 > gm12878_gex_1 ~ gm12878_gex_2
  ),
  k562_gex = case_when(
    is.na(k562_gex_1) ~ k562_gex_2,
    is.na(k562_gex_2) ~ k562_gex_1,
    k562_gex_1 > k562_gex_2 ~ k562_gex_1,
    k562_gex_2 > k562_gex_1 ~ k562_gex_2
  )) %>%
  mutate(gex_diff = gm12878_gex-k562_gex)
```

```{r}
boxplot_df <- loops_gex %>%
  mutate(decon_diff = diff_after-diff_before) %>%
  filter(!is.na(decon_diff), !is.na(gex_diff)) %>%
  mutate(decon_diff_bin = cut(decon_diff,breaks=c(-400,-50,-25,-10,10,25,50,400)))

label_df <- boxplot_df %>%
  group_by(decon_diff_bin) %>%
  summarise(count=n())
labels <- paste0(label_df$decon_diff_bin, '\nn=', prettyNum(label_df$count, big.mark = ','))

boxplot_df %>%
  ggplot(aes(x = decon_diff_bin, y = gex_diff)) +
  geom_boxplot() +
  geom_hline(yintercept=0, lty='dashed', color='red') +
  coord_cartesian(ylim=c(-50,50)) +
  theme_classic() +
  scale_x_discrete(breaks = label_df$decon_diff_bin, labels = labels) +
  xlab('Avg. loop strength difference (After - Before) deconvolution') +
  ylab('Avg. gene expression difference')
ggsave('h3k27ac_gm12878_k562_loop_vs_gex_boxplots.pdf', height=4, width=6)
```


# Compare whether or not loops would be significant before/after deconvolution using DESeq2

```{r}
library(DESeq2)
before_gex <- before[loops_gex$id,]
after_gex <- round(after[loops_gex$id,])
na_idx <- which(rowSums(is.na(before_gex))>0 | rowSums(is.na(after_gex))>0)
# before
cts_before <- before_gex[-na_idx,]
cts_after <- after_gex[-na_idx,]
loops_subset <- loops_gex[-na_idx,]
idx_filter_10 <- which(rowSums(cts_before)>=10 & rowSums(cts_after)>=10)
cts_before <- cts_before[idx_filter_10,]
cts_after <- cts_after[idx_filter_10,]
loops_subset <- loops_subset[idx_filter_10,]
loops_subset$id <- paste(loops_subset$label, loops_subset$chr1, loops_subset$idx, sep='-')
rownames(cts_before) <- rownames(cts_after) <- loops_subset$id
colnames(cts_before) <- c('gm12878_1','gm12878_2','k562_1','k562_2', 'k562_3')
coldata <- data.frame(condition = c('gm12878','gm12878','k562','k562', 'k562'))
rownames(coldata) <- colnames(cts_before)
dds <- DESeqDataSetFromMatrix(countData = cts_before, colData = coldata, design = ~ condition)
dds <- DESeq(dds)
res_before <- results(dds)
# after
colnames(cts_after) <- c('gm12878_1','gm12878_2','k562_1','k562_2', 'k562_3')
coldata <- data.frame(condition = c('gm12878','gm12878','k562','k562', 'k562'))
rownames(coldata) <- colnames(cts_after)
dds <- DESeqDataSetFromMatrix(countData = round(cts_after), colData = coldata, design = ~ condition)
dds <- DESeq(dds)
res_after <- results(dds)
```

```{r}
loops_subset %>%
  mutate(log2fc_before = res_before$log2FoldChange, log2fc_after = res_after$log2FoldChange) %>%
  ggplot(aes(x = log2fc_before, y = log2fc_after)) +
  geom_point(alpha=0.1) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  geom_hline(yintercept=-1, color='blue') +
  geom_vline(xintercept=-1, color='blue') +
  theme_classic()
```


```{r}
loops_subset %>%
  mutate(log2fc_before = res_before$log2FoldChange, log2fc_after = res_after$log2FoldChange) %>%
  mutate(log2fc_bin = cut_number(log2fc_before,5)) %>%
  ggplot(aes(x = log2fc_bin, y = -gex_diff)) +
  geom_boxplot() +
  coord_cartesian(ylim=c(-50,50)) +
  theme_classic()
```

```{r}
loops_subset %>%
  mutate(log2fc_before = res_before$log2FoldChange, log2fc_after = res_after$log2FoldChange) %>%
  mutate(log2fc_bin = cut(log2fc_after-log2fc_before,10)) %>%
  ggplot(aes(x = log2fc_bin, y = gex_diff)) +
  geom_boxplot() +
  geom_hline(yintercept=0, lty='dashed', color='red') +
  coord_cartesian(ylim=c(-50,50)) +
  theme_classic()
```


## Volcano plots before/after

```{r}
# before
data.frame(res_before) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(alpha=0.1) +
  theme_classic()
```


```{r}
# before
data.frame(res_after) %>%
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(alpha=0.1) +
  theme_classic()
```


```{r}
new_sig <- which(res_before$padj > 0.1 & res_after$padj < 0.001)
stay_sig <- which(res_before$padj<0.001 & res_after$padj < 0.001)
not_sig <- which(res_before$padj < 0.001 & res_after$padj > 0.1)
```


```{r}
loops_subset$label <- ''
loops_subset$label[new_sig] <- 'new'
loops_subset$label[stay_sig] <- 'stay'
loops_subset$label[not_sig] <- 'not'
```

```{r}
loops_subset %>%
  mutate(log2fc_before = res_before$log2FoldChange, log2fc_after = res_after$log2FoldChange) %>%
  filter(label!='', label != 'stay') %>%
  ggplot(aes(x = log2fc_after, y = gex_diff)) +
  facet_wrap(label ~ .) +
  geom_point()
```


```{r}
data.frame(log2fc_before = res_before$log2FoldChange,
           log2fc_after = res_after$log2FoldChange,
           gex_diff = -loops_subset$gex_diff) %>%
  pivot_longer(cols=c(log2fc_before, log2fc_after), names_to = 'type', values_to = 'log2fc') %>%
  mutate(log2fc_bin = cut(log2fc, 5)) %>%
  ggplot(aes(x = log2fc_bin, y = gex_diff)) +
  geom_boxplot(aes(fill = type)) +
  scale_color_tableau() +
  coord_cartesian(ylim=c(-50,50)) +
  theme_classic()
```








