---
title: "HiChIP loop strengthening"
output: html_document
---

```{r}
library(GenomicRanges)
library(InteractionSet)
library(data.table)
library(rtracklayer)
library(Matrix)
library(spatstat)
library(dplyr)
library(tidyr)
library(ggplot2)
library(viridis)
library(latex2exp)

library(foreach)
library(doSNOW)
source('/rafalab/lzou/dcon/R/utils.R')

#CORES=16
```


```{r}
# cl <- makeCluster(CORES, outfile="") # number of cores. Notice 'outfile'
# registerDoSNOW(cl)
```


```{r}
# create sparse matrix from hic data
mychroms <- c(paste0('chr',seq(1,22)),'chrX','chrY')
coords <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_17_2.5kb_MatrixHiCPro_abs.bed',
                col.names = c('chrom','start','end','id'))
chrom_starts <- coords %>% group_by(chrom) %>% dplyr::slice(1)
coords <- makeGRangesFromDataFrame(coords, keep.extra.columns=T)
# load in the matrix and 0 out the diagonal + a stripe around the diagonal of width 3
mat <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_17_2.5kb_MatrixHiCPro.matrix')
mat <- mat %>% filter(abs(V2-V1)>3)
mat <- sparseMatrix(i = mat$V1, j = mat$V2, x = mat$V3, dims=c(length(coords),length(coords)))
mat2 <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_18_2.5kb_MatrixHiCPro.matrix')
mat2 <- mat2 %>% filter(abs(V2-V1)>3)
mat2 <- sparseMatrix(i = mat2$V1, j = mat2$V2, x = mat2$V3, dims=c(length(coords),length(coords)))
```

```{r}
anchors <- import('/rafalab/lzou/hichip_mnase/K562_CTCF_chip_peaks_hg38_ENCFF651WYF.bigBed')
h3k27ac <- import('/rafalab/lzou/hichip_mnase/K562_H3K27ac_chip_peaks_hg38_ENCFF045OHM.bigBed')
hic <- readRDS('/rafalab/lzou/hichip_mnase/K562_HiC_chr22_2500.rds')
```

```{r}
# macs summits
summits1 <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_17_S1.mapped.PT_primary_align_macs2_summits.bed', header=F, col.names = c('chrom','start','end','name','score'))
summits2 <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_18_S2.mapped.PT.primary.align_summits.bed', header=F, col.names = c('chrom','start','end','name','score'))
```

```{r}
# fithichip
fithichip1 <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_17_S1_2.5kb.interactions_FitHiC.bed')
fithichip2 <- fread('/rafalab/lzou/hichip_mnase/Deep_HC_YET_18_S2_2.5kb.interactions_FitHiC.bed')
```

```{r}
plot_mat <- function(mat, main='') {
  print(
    reshape2::melt(as.matrix(mat)) %>%
      ggplot(aes(x = Var2, y = Var1)) +
      geom_tile(aes(fill=value),color='grey') +
      scale_fill_gradient(low='white',high='red') +
      theme_classic() +
      xlab('Anchor 1') +
      ylab('Anchor 2') + 
      annotate("rect", xmin = 9.5, xmax = 12.5, ymin = 9.5, ymax = 12.5, color='blue', lty='dashed', fill=NA) +
      ggtitle(main)
  )
  
}
# symmetrically pad on either side of each grange
pad_windows <- function(granges, window_size=5e4, bin_size=2500) {
  x <- (window_size/2)-floor(width(ranges(granges))/2)
  starts <- start(ranges(granges))-x
  ends <- end(ranges(granges))+x
  short_idx <- which((ends-starts)!=window_size)
  ends[short_idx] <- ends[short_idx]+1 
  # add one to the starts that are divisible so it doesn't pick up the previous
  starts_0 <- (starts%%bin_size)==0
  starts[starts_0] <- starts[starts_0]+1
  start(ranges(granges)) <- starts
  end(ranges(granges)) <- ends
  granges <- trim(granges)
  return(granges)
}
normalize_hic <- function(M, sigma=0.2, gamma=1) {
  M <- as.matrix(M)
  zero_idx <- which(rowSums(M)==0)
  if (length(zero_idx)>0) {
    for (id in zero_idx) {
      M[id,id] <- 1
    }
  }
  M <- M+t(M)-diag(diag(M))
  #Dsmooth <- as.matrix(blur(im(M),sigma=sigma))
  Dsmoothnorm <- sweep(M,2,colSums(M),'/')
  ILD <- (1-gamma)*diag(1,nrow=nrow(Dsmoothnorm))+gamma*Dsmoothnorm
  return(ILD)
}
fit_decon <- function(reads, D, DF=21) {
  B <- constructB(1:length(reads), df=DF)
  starting_a <- rep(0, DF)
  o <- optim(par=starting_a, loglik, y=reads, D=D, B=B, method='BFGS')
  a <- o$par
  # get hessian
  H <- hess_a(y=reads, D=D, B=B, a=a)
  Sigma <- solve(psd(H))
  BSigmaB <- B%*%Sigma%*%t(B)
  Bavars <- diag(BSigmaB)
  estimate <- as.numeric(exp(B%*%a))
  return(list(a=a, est=as.numeric(B%*%a), vars=Bavars))
}
decon_to_2d <- function(decon1, decon2, mat) {
  new_mat <- (decon1$est/sum(decon1$est))%*%(t(decon2$est)/sum(decon2$est))
  new_mat <- new_mat*sum(mat)
  return(new_mat)
}
```

```{r}
# sample some loops from chr22 and make anchor list
set.seed(1337)
f1 <- fithichip1 %>%
  filter(chr1=='chr22') %>%
  mutate(qval = -log10(`Q-Value_Bias`)) %>%
  mutate(qval_bin = cut(qval,breaks=c(0,1,5,10,500))) %>%
  group_by(qval_bin) %>%
  sample_n(2) %>%
  as.data.frame()
f1$bin1 <- left_join(f1[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin2 <- left_join(f1[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin_pair <- paste0(f1$bin1,',',f1$bin2)
f2 <- fithichip2 %>%
  filter(chr1=='chr22') %>%
  as.data.frame()
f2$bin1 <- left_join(f2[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin2 <- left_join(f2[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin_pair <- paste0(f2$bin1,',',f2$bin2)
f2 <- f2 %>% filter(bin_pair%in%f1$bin_pair)

f1$qval <- -log10(f1$`Q-Value_Bias`)
f2$qval <- -log10(f2$`Q-Value_Bias`)
```


```{r}
current_chrom <- 'chr22'
cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_',current_chrom,'_2500.mtx'))
```


```{r}
PAD=10
for (i in 1:nrow(f2)) {
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx])
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx])
  m1 <- mat[a1_idx,a2_idx]
  m2 <- mat2[a1_idx,a2_idx]
  if (nnzero(m1) > 0) {
    plot_mat(m1, main=paste0('Anchor pair ', i, ', Replicate 1: qval: ', f1$qval[f1idx]))
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    plot_mat(res1, main='Deconvolved')
  }
  if (nnzero(m2) > 0) {
    plot_mat(m2, main=paste0('Anchor pair ', i, ', Replicate 2: qval: ', f2$qval[i]))
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    plot_mat(res2, main='Deconvolved')
  }
}
```

```{r}
# do for all loops on chr22
f1 <- fithichip1 %>%
  filter(chr1=='chr22') %>%
  mutate(qval = -log10(`Q-Value_Bias`)) %>%
  as.data.frame()
f1$bin1 <- left_join(f1[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin2 <- left_join(f1[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin_pair <- paste0(f1$bin1,',',f1$bin2)
f2 <- fithichip2 %>%
  filter(chr1=='chr22') %>%
  as.data.frame()
f2$bin1 <- left_join(f2[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin2 <- left_join(f2[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin_pair <- paste0(f2$bin1,',',f2$bin2)
f2 <- f2 %>% filter(bin_pair%in%f1$bin_pair)

f1$qval <- -log10(f1$`Q-Value_Bias`)
f2$qval <- -log10(f2$`Q-Value_Bias`)
```

```{r}
PAD=10
results <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                      rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                      rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                      rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                      dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                      L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results$bin1[i] <- a1
  results$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  results$rep1_log10qval[i] <- f1$qval[f1idx]
  results$rep2_log10qval[i] <- f2$qval[i]
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- mat[a1_idx,a2_idx]
  m2 <- mat2[a1_idx,a2_idx]
  results$rep1_before[i] <- sum(m1[10:12,10:12])
  results$rep2_before[i] <- sum(m2[10:12,10:12])
  results$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results$L1_after[i] <- (sum(abs(res2-res1)))
}
```


```{r}
results %>%
  ggplot(aes(x = rep1_before, y = rep2_before)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  annotate('text', x=50, y = 200, label=paste0('r2=',round(cor(results$rep1_before, results$rep2_before)^2,3))) +
  theme_classic()
```


```{r}
results %>%
  ggplot(aes(x = rep1_after, y = rep2_after)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  annotate('text', x=50, y = 200, label=paste0('r2=',round(cor(results$rep1_after, results$rep2_after)^2,3))) +
  theme_classic()
```



```{r}
results %>%
  ggplot(aes(x = dist_before, y = dist_after)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  xlab('L2 distance before') +
  ylab('L2 distance after') +
  ggtitle('chr22 significant loops, rep 1 vs. rep 2') +
  theme_classic()
```



```{r}
results %>%
  ggplot(aes(x = L1_before, y = L1_after)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  xlab('L1 distance before') +
  ylab('L1 distance after') +
  ggtitle('chr22 significant loops, rep 1 vs. rep 2') +
  theme_classic()
```

```{r}
results[which(results$dist_before<40 & results$dist_after>500),]
```






```{r, eval=F}
a1 <- 666205
a2 <- 666249
f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
# pad by 10 on each side
a1_idx <- seq(a1-PAD,a1+PAD)
a2_idx <- seq(a2-PAD,a2+PAD)
hic1_idx <- a1_idx-cs+1
hic2_idx <- a2_idx-cs+1
hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.8)
hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=1)
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]
if (nnzero(m1) > 0) {
  plot_mat(m1, main=paste0('Anchor pair ', 0, ', Replicate 1: qval: ', f1$qval[f1idx]))
  r1 <- rowSums(m1)
  r2 <- colSums(m1)
  B <- constructB(1:length(r1), df=21)
  fit_r1 <- fit_decon(r1, hic1, DF=21)
  fit_r2 <- fit_decon(r2, hic2, DF=21)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
  plot_mat(res1, main='Deconvolved')
}
if (nnzero(m2) > 0) {
  plot_mat(m2, main=paste0('Anchor pair ', 0, ', Replicate 2: before'))
  r1 <- rowSums(m2)
  r2 <- colSums(m2)
  B <- constructB(1:length(r1), df=21)
  fit_r1 <- fit_decon(r1, hic1, DF=21)
  fit_r2 <- fit_decon(r2, hic2, DF=21)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
  plot_mat(res2, main='After; Deconvolved with gamma=1')
}
```



```{r, eval=F}
a1 <- 666205
a2 <- 666249
f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
# pad by 10 on each side
a1_idx <- seq(a1-PAD,a1+PAD)
a2_idx <- seq(a2-PAD,a2+PAD)
hic1_idx <- a1_idx-cs+1
hic2_idx <- a2_idx-cs+1
hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.8)
hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.8)
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]
if (nnzero(m1) > 0) {
  plot_mat(m1, main=paste0('Anchor pair ', 0, ', Replicate 1: qval: ', f1$qval[f1idx]))
  r1 <- rowSums(m1)
  r2 <- colSums(m1)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
  plot_mat(res1, main='Deconvolved')
}
if (nnzero(m2) > 0) {
  plot_mat(m2, main=paste0('Anchor pair ', 0, ', Replicate 2: before'))
  r1 <- rowSums(m2)
  r2 <- colSums(m2)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
  plot_mat(res2, main='After; Deconvolved with gamma=0.8')
}
```


# LPS data

```{r}
lps1 <- fread('/rafalab/lzou/hichip_lps/LPS853_1_allValidPairs_2500.matrix')
lps1 <- lps1 %>% filter(abs(V2-V1)>3)
lps1 <- sparseMatrix(i = lps1$V1, j = lps1$V2, x = lps1$V3, dims=c(length(coords),length(coords)))
lps2 <- fread('/rafalab/lzou/hichip_lps/LPS853_2_allValidPairs_2500.matrix')
lps2 <- lps2 %>% filter(abs(V2-V1)>3)
lps2 <- sparseMatrix(i = lps2$V1, j = lps2$V2, x = lps2$V3, dims=c(length(coords),length(coords)))
```

```{r}
lps11 <- fread('/rafalab/lzou/hichip_lps/LPS141_1_allValidPairs_2500.matrix')
lps11 <- lps11 %>% filter(abs(V2-V1)>3)
lps11 <- sparseMatrix(i = lps11$V1, j = lps11$V2, x = lps11$V3, dims=c(length(coords),length(coords)))
lps22 <- fread('/rafalab/lzou/hichip_lps/LPS141_2_allValidPairs_2500.matrix')
lps22 <- lps22 %>% filter(abs(V2-V1)>3)
lps22 <- sparseMatrix(i = lps22$V1, j = lps22$V2, x = lps22$V3, dims=c(length(coords),length(coords)))
```

## LPS853 

```{r}
# sample some loops from chr22 and make anchor list
set.seed(1337)
f1 <- fithichip1 %>%
  filter(chr1=='chr22') %>%
  mutate(qval = -log10(`Q-Value_Bias`)) %>%
  mutate(qval_bin = cut(qval,breaks=c(0,1,5,10,500))) %>%
  group_by(qval_bin) %>%
  sample_n(2) %>%
  as.data.frame()
f1$bin1 <- left_join(f1[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin2 <- left_join(f1[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin_pair <- paste0(f1$bin1,',',f1$bin2)
f2 <- fithichip2 %>%
  filter(chr1=='chr22') %>%
  as.data.frame()
f2$bin1 <- left_join(f2[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin2 <- left_join(f2[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin_pair <- paste0(f2$bin1,',',f2$bin2)
f2 <- f2 %>% filter(bin_pair%in%f1$bin_pair)

f1$qval <- -log10(f1$`Q-Value_Bias`)
f2$qval <- -log10(f2$`Q-Value_Bias`)
```

```{r}
PAD=10
for (i in 1:nrow(f2)) {
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- lps1[a1_idx,a2_idx]
  m2 <- lps2[a1_idx,a2_idx]
  if (nnzero(m1) > 0) {
    plot_mat(m1, main=paste0('Anchor pair ', i, ', Replicate 1: qval: ', f1$qval[f1idx]))
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    plot_mat(res1, main='Deconvolved')
  }
  if (nnzero(m2) > 0) {
    plot_mat(m2, main=paste0('Anchor pair ', i, ', Replicate 2: qval: ', f2$qval[i]))
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    plot_mat(res2, main='Deconvolved')
  }
}
```


```{r}
# do for all loops on chr22
f1 <- fithichip1 %>%
  filter(chr1=='chr22') %>%
  mutate(qval = -log10(`Q-Value_Bias`)) %>%
  as.data.frame()
f1$bin1 <- left_join(f1[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin2 <- left_join(f1[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f1$bin_pair <- paste0(f1$bin1,',',f1$bin2)
f2 <- fithichip2 %>%
  filter(chr1=='chr22') %>%
  as.data.frame()
f2$bin1 <- left_join(f2[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin2 <- left_join(f2[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id)) %>% pull(id)
f2$bin_pair <- paste0(f2$bin1,',',f2$bin2)
f2 <- f2 %>% filter(bin_pair%in%f1$bin_pair)

f1$qval <- -log10(f1$`Q-Value_Bias`)
f2$qval <- -log10(f2$`Q-Value_Bias`)
```

```{r}
PAD=10
results_lps <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                          rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                          rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                          rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                          dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                          L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results_lps$bin1[i] <- a1
  results_lps$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  results_lps$rep1_log10qval[i] <- f1$qval[f1idx]
  results_lps$rep2_log10qval[i] <- f2$qval[i]
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- lps1[a1_idx,a2_idx]
  m2 <- lps2[a1_idx,a2_idx]
  results_lps$rep1_before[i] <- sum(m1[10:12,10:12])
  results_lps$rep2_before[i] <- sum(m2[10:12,10:12])
  results_lps$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results_lps$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results_lps$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results_lps$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results_lps$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results_lps$L1_after[i] <- (sum(abs(res2-res1)))
}
```

```{r}
results_lps %>%
  filter(!is.na(rep1_after), !is.na(rep2_after)) %>%
  ggplot(aes(x = L1_before, y = L1_after)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  xlab('L1 distance before') +
  ylab('L1 distance after') +
  ggtitle('chr22 significant loops for K562 CTCF, LPS853 rep 1 vs. rep 2') +
  theme_classic()
```


## LPS141

```{r}
PAD=10
results_lps141 <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                             rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                             rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                             rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                             dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                             L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results_lps141$bin1[i] <- a1
  results_lps141$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  results_lps141$rep1_log10qval[i] <- f1$qval[f1idx]
  results_lps141$rep2_log10qval[i] <- f2$qval[i]
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- lps11[a1_idx,a2_idx]
  m2 <- lps22[a1_idx,a2_idx]
  results_lps141$rep1_before[i] <- sum(m1[10:12,10:12])
  results_lps141$rep2_before[i] <- sum(m2[10:12,10:12])
  results_lps141$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results_lps141$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results_lps141$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results_lps141$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results_lps141$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results_lps141$L1_after[i] <- (sum(abs(res2-res1)))
}
```



```{r}
results_lps141 %>%
  filter(!is.na(rep1_after), !is.na(rep2_after)) %>%
  ggplot(aes(x = L1_before, y = L1_after)) +
  geom_point(alpha=0.5) +
  geom_abline(slope=1, intercept=0, lty='dashed', color='red') +
  xlab('L1 distance before') +
  ylab('L1 distance after') +
  ggtitle('chr22 significant loops for K562 CTCF, LPS141 rep 1 vs. rep 2') +
  theme_classic()
```



# Compare LPS and K562 before and after


```{r}
before_mat <- rbind(results$rep1_before, results$rep2_before, results_lps$rep1_before, results_lps$rep2_before, results_lps141$rep1_before, results_lps141$rep2_before)
after_mat <- rbind(results$rep1_after, results$rep2_after, results_lps$rep1_after, results_lps$rep2_after, results_lps141$rep1_after, results_lps141$rep2_after)
```

```{r}
pr_before <- prcomp(before_mat, center=T, scale=T, rank=2)
before_mat <- before_mat[,-which(colSums(is.na(after_mat))>0)]
after_mat <- after_mat[,-which(colSums(is.na(after_mat))>0)]
pr_after <- prcomp(after_mat, center=T, scale=T, rank=2)
```

```{r}
labels <- c('K562', 'K562', 'LPS853', 'LPS853', 'LPS141', 'LPS141')
bind_rows(pr_before$x %>% as.data.frame() %>% mutate(label=labels, type=rep('before',6)),
          pr_after$x %>% as.data.frame() %>% mutate(label=labels, type=rep('after',6))) %>%
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(color=label)) +
  facet_wrap(type ~ .) +
  theme_bw() +
  xlab('PC1 73% (before) and 76% (after)') +
  ylab('PC2 10% (before) and 11% (after)')
```


```{r}
cor_before_mat <- cor(t(before_mat))^2
cor_after_mat <- cor(t(after_mat))^2
dimnames(cor_before_mat) <- dimnames(cor_after_mat) <- list(c('K562_1', 'K562_2', 'LPS853_1', 'LPS853_2', 'LPS141_1', 'LPS141_2'),c('K562_1', 'K562_2', 'LPS853_1', 'LPS853_2', 'LPS141_1', 'LPS141_2'))

reshape2::melt(cor_before_mat) %>%
  mutate(type='before') %>%
  bind_rows(reshape2::melt(cor_after_mat) %>% mutate(type='after')) %>%
  mutate(type = factor(type, levels=c('before','after'))) %>%
  ggplot(aes(x = Var1, y = Var2)) +
  geom_tile(aes(fill = value), color='black') +
  geom_text(aes(label=round(value,2))) +
  scale_fill_gradient(TeX(r"($r^2$)"),low='white', high='forestgreen',limits=c(0,1)) +
  facet_wrap(type ~ .) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  xlab('') +
  ylab('') +
  scale_y_discrete(limits=rev) +
  ggtitle('correlation pre- and post- deconvolution \nfor K562 chr22 50kb loop anchor windows')
```


## Figure out why correlation between K562 and LPS goes to 0

```{r}
data.frame(K562_1 = results$rep1_before, LPS853_1 = results_lps$rep1_before,
           type = 'before') %>%
  bind_rows(data.frame(K562_1 = results$rep1_after, LPS853_1 = results_lps$rep1_after,
                       type = 'after')) %>%
  mutate(type = factor(type, levels=c('before','after'))) %>%
  ggplot(aes(x = K562_1, y = LPS853_1)) +
  geom_point(alpha=0.25) +
  facet_wrap(type ~ .) +
  theme_classic()
```


```{r}
data.frame(K562_1 = results$rep1_before, K562_2 = results$rep2_before,
           type = 'before') %>%
  bind_rows(data.frame(K562_1 = results$rep1_after, K562_2 = results$rep2_after,
                       type = 'after')) %>%
  mutate(type = factor(type, levels=c('before','after'))) %>%
  ggplot(aes(x = K562_1, y = K562_2)) +
  geom_point(alpha=0.25) +
  facet_wrap(type ~ .) +
  theme_classic()
```

## TODO: LOG AND RMSE; CORRELATION WITH GENE EXPRESSION

# Compare on LPS141 loops

```{r}
# fithichip
RESOLUTION=2500
MIN_PET=3
lps141_loops1 <- bind_rows(fread('/rafalab/lzou/hichip_lps/loops/LPS141_1_E_P_loops.bedpe'), fread('/rafalab/lzou/hichip_lps/loops/LPS141_1_P_P_loops.bedpe'), fread('/rafalab/lzou/hichip_lps/loops/LPS141_1_E_E_loops.bedpe'))
lps141_loops2 <- bind_rows(fread('/rafalab/lzou/hichip_lps/loops/LPS141_2_E_P_loops.bedpe'), fread('/rafalab/lzou/hichip_lps/loops/LPS141_2_P_P_loops.bedpe'), fread('/rafalab/lzou/hichip_lps/loops/LPS141_2_E_E_loops.bedpe'))

colnames(lps141_loops1) <- colnames(lps141_loops2) <- c('chr1', 's1', 'e1', 'chr2', 's2', 'e2', 'idk', 'pet')

# TODO: FIGURE OUT WHICH BIN IT'S MOSTLY IN INSTEAD OF JUST ROUNDING THE START DOWN AND USING THAT

# only keep loops appearing in both replicates
# convert starts to 2.5kb bins
lps141_loops1$s1 <- floor(lps141_loops1$s1/RESOLUTION)*RESOLUTION
lps141_loops1$s2 <- floor(lps141_loops1$s2/RESOLUTION)*RESOLUTION
lps141_loops1$e1 <- floor(lps141_loops1$e1/RESOLUTION)*RESOLUTION
lps141_loops1$e2 <- floor(lps141_loops1$e2/RESOLUTION)*RESOLUTION
lps141_loops2$s1 <- floor(lps141_loops2$s1/RESOLUTION)*RESOLUTION
lps141_loops2$s2 <- floor(lps141_loops2$s2/RESOLUTION)*RESOLUTION
lps141_loops2$e1 <- floor(lps141_loops2$e1/RESOLUTION)*RESOLUTION
lps141_loops2$e2 <- floor(lps141_loops2$e2/RESOLUTION)*RESOLUTION

f1 <- lps141_loops1 %>%
  filter(chr1=='chr22', pet>=MIN_PET) %>%
  as.data.frame()
f1$bin1 <- left_join(f1[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id), by=c('seqnames','start')) %>% pull(id)
f1$bin2 <- left_join(f1[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id), by=c('seqnames','start')) %>% pull(id)
f1$bin_pair <- paste0(f1$bin1,',',f1$bin2)
f2 <- lps141_loops2 %>%
  filter(chr1=='chr22', pet>=MIN_PET) %>%
  as.data.frame()
f2$bin1 <- left_join(f2[c('chr1','s1','e1')] %>% dplyr::rename(seqnames=chr1, start=s1, end=e1),as.data.frame(coords) %>% select(seqnames, start, end, id), by=c('seqnames','start')) %>% pull(id)
f2$bin2 <- left_join(f2[c('chr2','s2','e2')] %>% dplyr::rename(seqnames=chr2, start=s2, end=e2),as.data.frame(coords) %>% select(seqnames, start, end, id), by=c('seqnames','start')) %>% pull(id)
f2$bin_pair <- paste0(f2$bin1,',',f2$bin2)
f2 <- f2 %>% filter(bin_pair%in%f1$bin_pair)
```

```{r}
PAD=10
results <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                      rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                      rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                      rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                      dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                      L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results$bin1[i] <- a1
  results$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- mat[a1_idx,a2_idx]
  m2 <- mat2[a1_idx,a2_idx]
  results$rep1_before[i] <- sum(m1[10:12,10:12])
  results$rep2_before[i] <- sum(m2[10:12,10:12])
  results$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results$L1_after[i] <- (sum(abs(res2-res1)))
}
```

```{r}
PAD=10
results_lps <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                          rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                          rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                          rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                          dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                          L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results_lps$bin1[i] <- a1
  results_lps$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- lps1[a1_idx,a2_idx]
  m2 <- lps2[a1_idx,a2_idx]
  results_lps$rep1_before[i] <- sum(m1[10:12,10:12])
  results_lps$rep2_before[i] <- sum(m2[10:12,10:12])
  results_lps$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results_lps$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results_lps$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results_lps$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results_lps$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results_lps$L1_after[i] <- (sum(abs(res2-res1)))
}
```

```{r}
PAD=10
results_lps141 <- data.frame(bin1 = numeric(nrow(f2)), bin2 = numeric(nrow(f2)),
                             rep1_log10qval = numeric(nrow(f2)), rep2_log10qval = numeric(nrow(f2)),
                             rep1_before = rep(NA,nrow(f2)), rep2_before = rep(NA,nrow(f2)),
                             rep1_after = rep(NA,nrow(f2)), rep2_after = rep(NA,nrow(f2)),
                             dist_before = rep(NA, nrow(f2)), dist_after = rep(NA, nrow(f2)),
                             L1_before = rep(NA, nrow(f2)), L1_after = rep(NA, nrow(f2)))
for (i in 1:nrow(f2)) {
  if (i%%10==0) {
    print(i)
  }
  a1 <- f2$bin1[i]
  a2 <- f2$bin2[i]
  results_lps141$bin1[i] <- a1
  results_lps141$bin2[i] <- a2
  f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
  # pad by 10 on each side
  a1_idx <- seq(a1-PAD,a1+PAD)
  a2_idx <- seq(a2-PAD,a2+PAD)
  hic1_idx <- a1_idx-cs+1
  hic2_idx <- a2_idx-cs+1
  hic1 <- normalize_hic(hic[hic1_idx,hic1_idx], gamma=0.5)
  hic2 <- normalize_hic(hic[hic2_idx,hic2_idx], gamma=0.5)
  m1 <- lps11[a1_idx,a2_idx]
  m2 <- lps22[a1_idx,a2_idx]
  results_lps141$rep1_before[i] <- sum(m1[10:12,10:12])
  results_lps141$rep2_before[i] <- sum(m2[10:12,10:12])
  results_lps141$dist_before[i] <- sqrt(sum((m2-m1)^2))
  results_lps141$L1_before[i] <- (sum(abs(m2-m1)))
  if (nnzero(m1) > 0) {
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
    results_lps141$rep1_after[i] <- sum(res1[10:12,10:12])
  }
  if (nnzero(m2) > 0) {
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    B <- constructB(1:length(r1), df=21)
    fit_r1 <- fit_decon(r1, hic1)
    fit_r2 <- fit_decon(r2, hic2)
    a1 <- fit_r1$a
    a2 <- fit_r2$a
    ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
    ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
    res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
    results_lps141$rep2_after[i] <- sum(res2[10:12,10:12])
  }
  results_lps141$dist_after[i] <- sqrt(sum((res2-res1)^2))
  results_lps141$L1_after[i] <- (sum(abs(res2-res1)))
}
```


```{r}
before_mat <- rbind(results$rep1_before, results$rep2_before, results_lps$rep1_before, results_lps$rep2_before, results_lps141$rep1_before, results_lps141$rep2_before)
after_mat <- rbind(results$rep1_after, results$rep2_after, results_lps$rep1_after, results_lps$rep2_after, results_lps141$rep1_after, results_lps141$rep2_after)
```

```{r}
pr_before <- prcomp(before_mat, center=T, scale=T, rank=2)
after_mat <- after_mat[,-which(colSums(is.na(after_mat))>0)]
pr_after <- prcomp(after_mat, center=T, scale=T, rank=2)
```

```{r}
labels <- c('K562', 'K562', 'LPS853', 'LPS853', 'LPS141', 'LPS141')
bind_rows(pr_before$x %>% as.data.frame() %>% mutate(label=labels, type=rep('before',6)),
          pr_after$x %>% as.data.frame() %>% mutate(label=labels, type=rep('after',6))) %>%
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(aes(color=label)) +
  facet_wrap(type ~ .) +
  theme_bw() +
  xlab('PC1 73% (before) and 76% (after)') +
  ylab('PC2 10% (before) and 11% (after)')
```

```{r}
cor_before_mat <- cor(t(before_mat))^2
cor_after_mat <- cor(t(after_mat))^2
dimnames(cor_before_mat) <- dimnames(cor_after_mat) <- list(c('K562_1', 'K562_2', 'LPS853_1', 'LPS853_2', 'LPS141_1', 'LPS141_2'),c('K562_1', 'K562_2', 'LPS853_1', 'LPS853_2', 'LPS141_1', 'LPS141_2'))

reshape2::melt(cor_before_mat) %>%
  mutate(type='before') %>%
  bind_rows(reshape2::melt(cor_after_mat) %>% mutate(type='after')) %>%
  mutate(type = factor(type, levels=c('before','after'))) %>%
  ggplot(aes(x = Var1, y = Var2)) +
  geom_tile(aes(fill = value), color='black') +
  geom_text(aes(label=round(value,2))) +
  scale_fill_gradient(TeX(r"($r^2$)"),low='white', high='forestgreen',limits=c(0,1)) +
  facet_wrap(type ~ .) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  xlab('') +
  ylab('') +X$
  scale_y_discrete(limits=rev) +
  ggtitle('correlation pre- and post- deconvolution \nfor LPS141 chr22 50kb loop anchor windows')
```



```{r}
results <- results %>%
  mutate(log2raw1 = log2(rep1_before+1), log2raw2 = log2(rep2_before+1),
         log2new1 = log2(rep1_after+1), log2new2 = log2(rep2_after+1)) %>%
  mutate(M_before = log2raw1-log2raw2, M_after = log2new1-log2new2,
         A_before = .5*(log2raw1+log2raw2), A_after = .5*(log2new1+log2new2)) 
```

```{r}
dd <- rbind(results %>% select(M_before,A_before) %>% dplyr::rename(M=M_before,A=A_before) %>% mutate(type='before'), results %>% select(M_after,A_after) %>% dplyr::rename(M=M_after,A=A_after) %>% mutate(type='after'))

dd$qval1 <- rep(results$rep1_log10qval, times=2)
dd$qval2 <- rep(results$rep2_log10qval, times=2)

dd %>%
  mutate(avg_qval = (qval1+qval2)/2) %>%
  mutate(avg_qval_bin = cut(avg_qval, breaks=c(0,0.5,1,2,3,4,5,10,50,150))) %>%
  ggplot(aes(x = A, y = M)) +
  geom_point(aes(color=avg_qval_bin),alpha=0.2) +
  facet_wrap(type ~ .) +
  theme_classic() +
  geom_hline(yintercept=0, lty='dashed', col='blue')
```


```{r}
dd %>%
  mutate(min_qval = ifelse(qval1<qval2, qval1, qval2)) %>%
  mutate(min_qval_bin = cut(min_qval, breaks=c(0,0.5,1,2,3,4,5,10,50,150))) %>%
  ggplot(aes(x = A, y = M)) +
  geom_point(aes(color=min_qval_bin),alpha=0.2) +
  facet_wrap(type ~ .) +
  theme_classic() +
  geom_hline(yintercept=0, lty='dashed', col='blue')
```




```{r}
dd %>%
  mutate(avg_qval = (qval1+qval2)/2) %>%
  mutate(avg_qval_bin = cut(avg_qval, breaks=c(0,0.5,1,2,3,4,5,10,50,150))) %>%
  filter(avg_qval > 0.5) %>%
  ggplot(aes(x = A, y = M)) +
  geom_point(aes(color=avg_qval_bin),alpha=0.2) +
  facet_wrap(type ~ .) +
  theme_classic() +
  geom_hline(yintercept=0, lty='dashed', col='blue')
```


```{r}
results %>%
  mutate(avg_qval = (rep1_log10qval+rep2_log10qval)/2) %>%
  mutate(avg_qval_bin = cut(avg_qval, breaks=c(0,0.1,0.5,1,2,3,4,5,10,50,150))) %>%
  ggplot(aes(x=A_before, y=A_after)) +
  geom_point(aes(color=avg_qval_bin)) +
  geom_abline(slope=1,intercept=0,lty='dashed') +
  theme_classic()
```


```{r}
results %>% 
  filter(abs(M_before)<2, abs(M_after)>2)
```


```{r}
a1 <- 666123
a2 <- 666143
f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
# pad by 10 on each side
a1_idx <- seq(a1-PAD,a1+PAD)
a2_idx <- seq(a2-PAD,a2+PAD)
hic1_idx <- a1_idx-cs+1
hic2_idx <- a2_idx-cs+1
hic1 <- normalize_hic(hic[hic1_idx,hic1_idx])
hic2 <- normalize_hic(hic[hic2_idx,hic2_idx])
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]
if (nnzero(m1) > 0) {
  plot_mat(m1, main=paste0('Anchor pair ', 0, ', Replicate 1: qval: ', f1$qval[f1idx]))
  r1 <- rowSums(m1)
  r2 <- colSums(m1)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
  plot_mat(res1, main='Deconvolved')
}
if (nnzero(m2) > 0) {
  plot_mat(m2, main=paste0('Anchor pair ', 0, ', Replicate 2: qval: ', f2$qval[i]))
  r1 <- rowSums(m2)
  r2 <- colSums(m2)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
  plot_mat(res2, main='Deconvolved')
}
```


```{r}
a1 <- 666306
a2 <- 666317
f1idx <- which(f1$bin1==a1 & f1$bin2==a2)
# pad by 10 on each side
a1_idx <- seq(a1-PAD,a1+PAD)
a2_idx <- seq(a2-PAD,a2+PAD)
hic1_idx <- a1_idx-cs+1
hic2_idx <- a2_idx-cs+1
hic1 <- normalize_hic(hic[hic1_idx,hic1_idx])
hic2 <- normalize_hic(hic[hic2_idx,hic2_idx])
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]
if (nnzero(m1) > 0) {
  plot_mat(m1, main=paste0('Anchor pair ', 0, ', Replicate 1: qval: ', f1$qval[f1idx]))
  r1 <- rowSums(m1)
  r2 <- colSums(m1)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res1 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m1)
  plot_mat(res1, main='Deconvolved')
}
if (nnzero(m2) > 0) {
  plot_mat(m2, main=paste0('Anchor pair ', 0, ', Replicate 2: qval: ', f2$qval[i]))
  r1 <- rowSums(m2)
  r2 <- colSums(m2)
  B <- constructB(1:length(r1), df=20)
  fit_r1 <- fit_decon(r1, hic1, DF=20)
  fit_r2 <- fit_decon(r2, hic2, DF=20)
  a1 <- fit_r1$a
  a2 <- fit_r2$a
  ff1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
  ff2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
  res2 <- exp(matrix(data = rep(ff1, 21), nrow=21, ncol=21) + t(matrix(data = rep(ff2, 21), nrow = 21, ncol = 21)))*sum(m2)
  plot_mat(res2, main='Deconvolved')
} # CHECK IF CTCF MOTIF OR CTCF CHIP-SEQ PEAK; OFFSET OF 5? 
```



```{r}
# Get anchors which are also in summits
anchors <- anchors[seqnames(anchors)%in%mychroms]
summits_combined <- makeGRangesFromDataFrame(rbind(summits1, summits2), keep.extra.columns = T)
summits_combined <- summits_combined[seqnames(summits_combined)%in%mychroms]
ovl <- findOverlaps(anchors, summits_combined)
anchors <- anchors[unique(queryHits(ovl))]
# keep the avg score from summits
ovl <- findOverlaps(anchors, summits_combined)
score_match <- data.frame(qh = queryHits(ovl), sh = subjectHits(ovl), 
                          score = summits_combined$score[subjectHits(ovl)])
score_match <- score_match %>% group_by(qh) %>% summarise(score = mean(score))
anchors$score <- score_match$score
```

```{r}
# pick some on chromosome 22 that have a range of average scores
set.seed(1337)
anchors_22 <-anchors[seqnames(anchors)=='chr22']
anchors_22$score_bins <- cut(anchors_22$score, 5)
idx <- anchors_22@elementMetadata %>%
  as.data.frame() %>%
  mutate(idx = seq(1,length(anchors_22))) %>%
  group_by(score_bins) %>%
  sample_n(2) %>%
  pull(idx)
sampled <- anchors_22[idx]
sampled <- sort(sampled)
sampled <- pad_windows(sampled, window_size=5e4)
```

```{r}
# grab the hic matrices
anchors_hic <- list()
current_chrom <- 'chr22'
cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_',current_chrom,'_2500.mtx'))
ovl <- data.frame(findOverlaps(sampled, coords))
# all anchors should have a window of 21 overlaps now
# First pre-process all of the hi-c and put it into a list
ovl$chrom <- as.character(seqnames(sampled[ovl$queryHits,]))
for (i in 1:length(sampled)) {
  print(i)
  o <- ovl[ovl$queryHits==i,]
  o$subjectHits <- o$subjectHits-cs+1
  anchors_hic[[as.character(i)]] <- normalize_hic(hic[o$subjectHits,o$subjectHits])
}
```

```{r}
for (i in 1:length(sampled)) {
  print(i)
  for (j in 1:length(sampled)) {
    if (j > i) {
      a1 <- sampled[i]
      a2 <- sampled[j]
      a1_idx <- ovl$subjectHits[which(ovl$queryHits==i)]
      a2_idx <- ovl$subjectHits[which(ovl$queryHits==j)]
      m1 <- mat[a1_idx,a2_idx]
      m2 <- mat2[a1_idx,a2_idx]
      
      if (nnzero(m1) > 0) {
        plot_mat(m1, main=paste0('Anchors ,', i, ',', j, ' Replicate 1: Avg score: ', sampled[i]$score))
        r1 <- rowSums(m1)
        r2 <- colSums(m1)
        B <- constructB(1:length(r1), df=10)
        fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
        fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
        a1 <- fit_r1$a
        a2 <- fit_r2$a
        
        f1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
        f2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
        # point estimate
        res <- matrix(data = rep(f1, 21), nrow=21, ncol=21) + t(matrix(data = rep(f2, 21), nrow = 21, ncol = 21))
      }
      if (nnzero(m2) > 0) {
        plot_mat(m2, main=paste0('Anchors ,', i, ',', j, ' Replicate 2: Avg score: ', sampled[j]$score))
      }
      
      # 
      # r1 <- rowSums(m2)
      # r2 <- colSums(m2)
      # B <- constructB(1:length(r1), df=10)
      # fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
      # fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
      # a1 <- fit_r1$a
      # a2 <- fit_r2$a
      # 
      # f1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
      # f2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
      # # point estimate
      # res <- matrix(data = rep(f1, 21), nrow=21, ncol=21) + t(matrix(data = rep(f2, 21), nrow = 21, ncol = 21))
      # # variance at each point in the matrix
      # H1 <- hess_a(y=r1, D=anchors_hic[[adx1]], B=B, a=a1)
      # Sigma <- solve(psd(H1))
      # BSigmaB <- B%*%Sigma%*%t(B)
      # v1 <- diag(BSigmaB)
      # H2 <- hess_a(y=r2, D=anchors_hic[[adx2]], B=B, a=a2)
      # Sigma <- solve(psd(H2))
      # BSigmaB <- B%*%Sigma%*%t(B)
      # v2 <- diag(BSigmaB)
      # res_var <- matrix(data = rep(v1, 21), nrow=21, ncol=21) + t(matrix(data = rep(v2, 21), nrow = 21, ncol = 21))
    }
  }
}
```











```{r}

decon_anchors_chrom <- function(anchors, hichip_rep1, hichip_rep2, chrom_starts, chrom='chr22', thresh=1e-1) {
  hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_',chrom,'_2500.mtx'))
  chrom_start <- as.numeric(chrom_starts[chrom_starts$chrom==chrom,'id'])
  anchs <- anchors[seqnames(anchors)==chrom]
  anchs <- pad_windows(anchs, window_size=5e4)
  pb <- txtProgressBar(min = 1, max = length(anchs)-1, style = 3)
  results <- foreach(
    # i = 1:(length(anchs)-1),
    i = 1:3,
    .combine = 'rbind',
    .export = c('Matrix', 'nnzero')
  ) %dopar% {
    setTxtProgressBar(pb, i) 
    jidx <- (i+1):length(anchs)
    result_mat <- matrix(data=0, nrow=length(jidx), ncol = 4)
    for (j in 1:length(jidx)) {
      a1 <- anchs[i]
      a2 <- anchs[j]
      if (length(findOverlaps(a1,a2))>0) {
        next
      }
      res1 <- decon_anchor_pair(a1, a2, hichip_rep1, hic, chrom_start)
      res2 <- decon_anchor_pair(a1, a2, hichip_rep2, hic, chrom_start)
      res1[res1<thresh] <- 0
      res2[res2<thresh] <- 0
      result_mat[j,1:2] <- res1
      result_mat[j,3:4] <- res2
    }
    result_mat <- result_mat[-which(rowSums(result_mat)==0),]
    return(result_mat)
  }
}

decon_anchor_pair <- function(a1, a2, hichip, hic, chrom_start) {
  chrom <- as.character(seqnames(a1))
  hichip_idx1 <- coords[subjectHits(findOverlaps(a1, coords))]$id
  hichip_idx2 <- coords[subjectHits(findOverlaps(a2, coords))]$id
  hic_idx1 <- hichip_idx1-chrom_start+1 # thx R 1-based indexing
  hic_idx2 <- hichip_idx2-chrom_start+1
  m <- hichip[hichip_idx1,hichip_idx2]
  if (nnzero(m)>0) {
    h1 <- hic[h1_idx,h1_idx]
    h2 <- hic[h2_idx,h2_idx]
    r1 <- rowSums(m)
    r2 <- colSums(m)
    h1_norm <- normalize_hic(h1)
    h2_norm <- normalize_hic(h2)
    fit_r1 <- fit_decon(r1, h1_norm)
    fit_r2 <- fit_decon(r2, h2_norm)
    new_mat <- decon_to_2d(fit_r1, fit_r2, m)
    return(c(sum(m[10:12,10:12]),sum(new_mat[10:12,10:12])))
  } else {
    return(c(0,0))
  }
}
deconvolve_anchor_pair_viz <- function(a1, a2, loc1, loc2, mat, label='') {
  a1_idx <- coords[subjectHits(findOverlaps(a1, coords))]$id
  a2_idx <- coords[subjectHits(findOverlaps(a2, coords))]$id
  h1_idx <- a1_idx-658135+1
  h2_idx <- a2_idx-658135+1
  m <- mat[a1_idx,a2_idx]
  if (nnzero(m)>0) {
    h1 <- hic[h1_idx,h1_idx]
    h2 <- hic[h2_idx,h2_idx]
    r1 <- rowSums(m)
    r2 <- colSums(m)
    h1_norm <- normalize_hic(h1)
    h2_norm <- normalize_hic(h2)
    fit_r1 <- fit_decon(r1, h1_norm)
    fit_r2 <- fit_decon(r2, h2_norm)
    xlabs <- h1_idx*2500
    plot(xlabs,fit_r1$est,col='white',main='Anchor 1')
    abline(v=loc1,col='red',lty='dashed')
    lines(xlabs,fit_r1$est,col='blue')
    points(xlabs,r1)
    xlabs <- h2_idx*2500
    plot(xlabs,fit_r2$est,col='white',main='Anchor 2')
    abline(v=loc2,col='red',lty='dashed')
    lines(xlabs,fit_r2$est,col='blue')
    points(xlabs,r2)
    new_mat <- decon_to_2d(fit_r1, fit_r2, m)
    m <- as.matrix(m)
    print(
      m %>%
        reshape2::melt() %>%
        ggplot(aes(x=Var1,y=Var2)) +
        geom_tile(aes(fill=value),color='black') +
        scale_fill_continuous(low='white',high='black') +
        theme_classic() +
        ggtitle(sum(m[10:12,10:12]))
    )
    print(
      new_mat %>%
        reshape2::melt() %>%
        ggplot(aes(x=Var1,y=Var2)) +
        geom_tile(aes(fill=value),color='black') +
        scale_fill_continuous(low='white',high='black') +
        theme_classic() +
        ggtitle(sum(new_mat[10:12,10:12]))
    )
  }
}
```




```{r}
anchors <- import('/rafalab/lzou/hichip_mnase/K562_CTCF_chip_peaks_hg38_ENCFF651WYF.bigBed')
anchors <- anchors[seqnames(anchors)%in%mychroms]
anchors <- pad_windows(anchors,5e4)
ovl <- data.frame(findOverlaps(anchors, coords))
# remove all the ones that are cut off at the end of chromosome
ovl_count <- ovl %>% group_by(queryHits) %>% summarise(count=n())
anchors <- anchors[-ovl_count$queryHits[which(ovl_count$count<21)]]
ovl <- data.frame(findOverlaps(anchors, coords))
# all anchors should have a window of 21 overlaps now
# First pre-process all of the hi-c and put it into a list
ovl$chrom <- as.character(seqnames(anchors[ovl$queryHits,]))
ovl <- left_join(ovl, chrom_starts, by='chrom')
anchors_hic <- list()
current_chrom <- 'chr22'
cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_','chr22','_2500.mtx'))
pb <- txtProgressBar(min = 1, max = length(unique(ovl$queryHits)), style = 3)
anchors_hic <- foreach(
  i = unique(ovl$queryHits[which(ovl$chrom=='chr22')]),
  .combine = 'append',
  .export = c('readMM')
) %dopar% {
  setTxtProgressBar(pb, i) 
  o <- ovl[ovl$queryHits==i,]
  if (o$chrom[1]!=current_chrom) {
    current_chrom <- o$chrom[1]
    hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_',current_chrom,'_2500.mtx'))
    cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
  }
  o$subjectHits <- o$subjectHits-cs+1
  l <- list()
  l[[as.character(i)]] <- normalize_hic(hic[o$subjectHits,o$subjectHits])
  return(l)
}

```





```{r}
# Test out getting the distribution of functions 
adx1 <- 81
adx2 <- 91
a1 <- anchors[chr22_anchors_idx[adx1]]
a2 <- anchors[chr22_anchors_idx[adx2]]
if (start(a2)<end(a1)) {
  next
}
a1_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx1])]
a2_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx2])]
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]

r1 <- rowSums(m2)
r2 <- colSums(m2)
B <- constructB(1:length(r1), df=10)
fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
a1 <- fit_r1$a
a2 <- fit_r2$a

f1 <- (B%*%a1)-log(sum(exp(B%*%a1)))
f2 <- (B%*%a2)-log(sum(exp(B%*%a2)))
# point estimate
res <- matrix(data = rep(f1, 21), nrow=21, ncol=21) + t(matrix(data = rep(f2, 21), nrow = 21, ncol = 21))
# variance at each point in the matrix
H1 <- hess_a(y=r1, D=anchors_hic[[adx1]], B=B, a=a1)
Sigma <- solve(psd(H1))
BSigmaB <- B%*%Sigma%*%t(B)
v1 <- diag(BSigmaB)
H2 <- hess_a(y=r2, D=anchors_hic[[adx2]], B=B, a=a2)
Sigma <- solve(psd(H2))
BSigmaB <- B%*%Sigma%*%t(B)
v2 <- diag(BSigmaB)
res_var <- matrix(data = rep(v1, 21), nrow=21, ncol=21) + t(matrix(data = rep(v2, 21), nrow = 21, ncol = 21))

if (nnzero(m1)>0) {
  results$raw_count_rep1[i] <- sum(m1[10:12,10:12])
  
  
  if (is(fit_r1, 'try-error')) {
    next
  }
  
  if (is(fit_r2, 'try-error')) {
    next
  }
  new_mat <- decon_to_2d(fit_r1, fit_r2, m1)
  results$new_count_rep1[i] <- sum(new_mat[10:12,10:12])
}
if (nnzero(m2)>0) {
  results$raw_count_rep2[i] <- sum(m2[10:12,10:12])
  r1 <- rowSums(m2)
  r2 <- colSums(m2)
  fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
  if (is(fit_r1, 'try-error')) {
    next
  }
  fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
  if (is(fit_r2, 'try-error')) {
    next
  }
  new_mat <- decon_to_2d(fit_r1, fit_r2, m2)
  results$new_count_rep2[i] <- sum(new_mat[10:12,10:12])
}
```




```{r}
anchors <- import('/rafalab/lzou/hichip_mnase/K562_CTCF_chip_peaks_hg38_ENCFF651WYF.bigBed')
anchors <- anchors[seqnames(anchors)%in%mychroms]
anchors <- pad_windows(anchors,5e4)
ovl <- data.frame(findOverlaps(anchors, coords))
# remove all the ones that are cut off at the end of chromosome
ovl_count <- ovl %>% group_by(queryHits) %>% summarise(count=n())
anchors <- anchors[-ovl_count$queryHits[which(ovl_count$count<21)]]
ovl <- data.frame(findOverlaps(anchors, coords))
# all anchors should have a window of 21 overlaps now
# First pre-process all of the hi-c and put it into a list
ovl$chrom <- as.character(seqnames(anchors[ovl$queryHits,]))
ovl <- left_join(ovl, chrom_starts, by='chrom')
anchors_hic <- list()
current_chrom <- 'chr22'
cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_','chr22','_2500.mtx'))
pb <- txtProgressBar(min = 1, max = length(unique(ovl$queryHits)), style = 3)
anchors_hic <- foreach(
  i = unique(ovl$queryHits[which(ovl$chrom=='chr22')]),
  .combine = 'append',
  .export = c('readMM')
) %dopar% {
  setTxtProgressBar(pb, i) 
  o <- ovl[ovl$queryHits==i,]
  if (o$chrom[1]!=current_chrom) {
    current_chrom <- o$chrom[1]
    hic <- readMM(paste0('/rafalab/lzou/hichip_mnase/K562_HiC_',current_chrom,'_2500.mtx'))
    cs <- chrom_starts$id[which(chrom_starts$chrom==current_chrom)]
  }
  o$subjectHits <- o$subjectHits-cs+1
  l <- list()
  l[[as.character(i)]] <- normalize_hic(hic[o$subjectHits,o$subjectHits])
  return(l)
}

# decon them all
chr22_anchors_idx <- which(seqnames(anchors)=='chr22')
results <- data.frame(anchor1 = rep(1:length(chr22_anchors_idx),each=length(chr22_anchors_idx)),
                      anchor2 = rep(1:length(chr22_anchors_idx),times=length(chr22_anchors_idx)),
                      raw_count_rep1 = numeric(length(chr22_anchors_idx)^2), 
                      raw_count_rep2 = numeric(length(chr22_anchors_idx)^2),
                      new_count_rep1 = numeric(length(chr22_anchors_idx)^2), 
                      new_count_rep2 = numeric(length(chr22_anchors_idx)^2))
results <- results[-which(results$anchor2<results$anchor1),]
results$raw_count_rep1 <- results$raw_count_rep2 <- results$new_count_rep1 <- results$new_count_rep2 <- NA
for (i in 242679:nrow(results)) {
  if(i%%1000==0) {print(i)}
  adx1 <- results$anchor1[i]
  adx2 <- results$anchor2[i]
  a1 <- anchors[chr22_anchors_idx[adx1]]
  a2 <- anchors[chr22_anchors_idx[adx2]]
  if (start(a2)<end(a1)) {
    next
  }
  a1_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx1])]
  a2_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx2])]
  m1 <- mat[a1_idx,a2_idx]
  m2 <- mat2[a1_idx,a2_idx]
  if (nnzero(m1)>0) {
    results$raw_count_rep1[i] <- sum(m1[10:12,10:12])
    r1 <- rowSums(m1)
    r2 <- colSums(m1)
    fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
    if (is(fit_r1, 'try-error')) {
      next
    }
    fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
    if (is(fit_r2, 'try-error')) {
      next
    }
    new_mat <- decon_to_2d(fit_r1, fit_r2, m1)
    results$new_count_rep1[i] <- sum(new_mat[10:12,10:12])
  }
  if (nnzero(m2)>0) {
    results$raw_count_rep2[i] <- sum(m2[10:12,10:12])
    r1 <- rowSums(m2)
    r2 <- colSums(m2)
    fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
    if (is(fit_r1, 'try-error')) {
      next
    }
    fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
    if (is(fit_r2, 'try-error')) {
      next
    }
    new_mat <- decon_to_2d(fit_r1, fit_r2, m2)
    results$new_count_rep2[i] <- sum(new_mat[10:12,10:12])
  }
}


saveRDS(results,file='loopstrength-results-200k-550k.rds')
```



```{r}
results <- readRDS('loopstrength-results-200k-550k.rds')

d <- results %>%
  filter(!is.na(raw_count_rep1) & !is.na(raw_count_rep2) & !is.na(new_count_rep1) & !is.na(new_count_rep2)) %>%
  mutate(log2raw1 = log2(raw_count_rep1+1), log2raw2 = log2(raw_count_rep2+1),
         log2new1 = log2(new_count_rep1+1), log2new2 = log2(new_count_rep2+1)) %>%
  mutate(M_before = log2raw1-log2raw2, M_after = log2new1-log2new2,
         A_before = .5*(log2raw1+log2raw2), A_after = .5*(log2new1+log2new2)) 

dd <- rbind(d %>% select(M_before,A_before) %>% rename(M=M_before,A=A_before) %>% mutate(type='before'), d %>% select(M_after,A_after) %>% rename(M=M_after,A=A_after) %>% mutate(type='after'))

dd %>%
  ggplot(aes(x = A, y = M)) +
  geom_point(alpha=0.2) +
  facet_wrap(type ~ .) +
  theme_classic() +
  geom_hline(yintercept=0, lty='dashed', col='blue')
```

```{r}
# try  just taking max of raw_count and new_count
pseudo <- d %>%
  group_by(anchor1, anchor2) %>%
  mutate(count1=max(raw_count_rep1, new_count_rep1), count2=max(raw_count_rep2,new_count_rep2)) %>%
  select(anchor1,anchor2,count1,count2) %>%
  mutate(log2c1 = log2(count1+1), log2c2=log2(count2+1))%>%
  mutate(M = log2c1-log2c2, A=0.5*(log2c1+log2c2)) 

pseudo$type <- 'pseudo'

dd %>%
  rbind(pseudo %>% ungroup() %>% select(M,A,type)) %>%
  ggplot(aes(x = A, y = M)) +
  geom_point() +
  facet_wrap(type ~ .) +
  theme_classic() +
  geom_hline(yintercept=0, lty='dashed', col='blue')
```



```{r}
d %>%
  ggplot(aes(x = A_before, y = A_after)) +
  geom_point(aes(),alpha=0.1) +
  theme_classic() +
  geom_abline(slope=1,intercept=0,lty='dashed',col='red') +
  coord_cartesian(xlim=c(0,8), ylim=c(0,8))
```

```{r}
length(which(d$A_before>d$A_after))
```

```{r}
length(which(d$A_after>d$A_before))
```


```{r}
d[which(d$A_before>2 & d$A_after<2),]
```


```{r}
d[which(abs(d$M_before)<2 & abs(d$M_after)>2),]
```

```{r}
d[which(abs(d$M_after)<2 & abs(d$M_before)>2),]
```


```{r}
adx1 <- 81
adx2 <- 91
a1 <- anchors[chr22_anchors_idx[adx1]]
a2 <- anchors[chr22_anchors_idx[adx2]]
if (start(a2)<end(a1)) {
  next
}
a1_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx1])]
a2_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx2])]
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]

r1 <- rowSums(m1)
r2 <- colSums(m1)


fit_r1 <- try(fit_decon(r1, anchors_hic[[adx1]]), silent=T)
fit_r2 <- try(fit_decon(r2, anchors_hic[[adx2]]), silent=T)
plot(1:21, r1, ylim=c(0,25), main='Rep 1, Anchor 1')
lines(1:21, exp(fit_r1$est), col='blue')
plot(1:21, r2, ylim=c(0,25), main='Rep 1, Anchor 2')
lines(1:21, exp(fit_r2$est), col='blue')



if (is(fit_r2, 'try-error')) {
  next
}
new_mat
```

```{r}
m2
```


```{r}
d[which(d$A_before>6 & d$A_after==0),]
```

```{r}
adx1 <- 341
adx2 <- 356
a1 <- anchors[chr22_anchors_idx[adx1]]
a2 <- anchors[chr22_anchors_idx[adx2]]
if (start(a2)<end(a1)) {
  next
}
a1_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx1])]
a2_idx <- ovl$subjectHits[which(ovl$queryHits==chr22_anchors_idx[adx2])]
m1 <- mat[a1_idx,a2_idx]
m2 <- mat2[a1_idx,a2_idx]

m1
```




```{r}
results %>%
  ggplot(aes(x = raw_count_rep1, y = raw_count_rep2)) +
  geom_point() +
  theme_classic()
```

```{r}
results %>%
  ggplot(aes(x = new_count_rep1, y = new_count_rep2)) +
  geom_point() +
  theme_classic()
```


